NextJS App Router RSC sample (#6083)
AzureAD/microsoft-authentication-library-for-js: samples/msal-node-samples/nextjs-rsc-ts/README.md

This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).  
This sample assumes some previous knowledge with React Server Components and Next.js.

## What this sample showcases

### React Server Components

The example `AuthProvider` implementation, uses functions wrapped in `cache` from `react`. This allows you to do multiple calls in separate server components, without the functions being invoked more than 1 time per request.

### Optional auth

The root `src/app/page.tsx` simply renders different content based on the existence of an account from the `AuthProvider`.

### Forced auth

In `src/app/forced/layout.tsx` you'll see these lines of code;

```tsx
const { account } = await authProvider.authenticate();

if (!account) {
  redirect(await authProvider.getAuthCodeUrl(codeRequest, getCurrentUrl()));
}
```

This will redirect the user to azure at the earliest possible time, using HTTP headers on the initial request to the page.

This can be useful if your whole app relies on being authenticated.

> ### Gradual consent

Reviewer: ```suggestion
### Incremental consent
```

- `./src/app/profile/page.tsx`

Shows how to get data from graph in a server component.

- `./src/app/event/page.tsx`

Also shows how to get data from graph, but needs additional consent before it can use the API.

### AuthProvider

- `./src/utils/AuthProvider.ts`

Contains an implementation wrapping the usage of a ConfidentialClientApplication, that accepts an partition manager and a cache client for the [DistributedCachePlugin](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/caching.md#web-apps) that will be initialized internally. 

- `./src/services/auth.ts`

Initializes an instance of the AuthProvider, and provides it with a `RedisCacheClient` and a `SessionPartitionManager`.

## Running the sample

Create an App Registration in the Azure portal with the following:

- Redirect URI (**WEB**): http://localhost:3000/auth/callback
- API permissions `User.Read` and `Calendars.Read`
- A client secret

Set up a `.env.local` file with the following values

```env
AZURE_AD_CLIENT_ID=YOUR_CLIENT_ID
AZURE_AD_CLIENT_SECRET=YOUR_CLIENT_SECRET

> AZURE_AD_TENANT_ID=YOUR_TENANT_ID

Reviewer: ```suggestion
AZURE_AD_AUTHORITY=YOUR_AZURE_AD_AUTHORITY
```

Reviewer: per serverConfig.ts

> # @azure/msal-node for Next.js App Router example

Reviewer 2: ```suggestion
# @azure/msal-node for Next.js with Server Components
```

### Server actions and Route Handlers

The `AuthProvider` works with Server Actions and Route Handlers, but the cache won't work the same way as with react. Therefore, a single call to `AuthProvider.authenticate` is recommended.

## Running the sample

Create an App Registration in the Azure portal with the following:

- Redirect URI (**WEB**): http://localhost:3000/auth/callback
- API permissions `User.Read` and `Calendars.Read`
- A client secret

Set up a `.env.local` file with the following values

```env
AZURE_AD_CLIENT_ID=YOUR_CLIENT_ID
AZURE_AD_CLIENT_SECRET=YOUR_CLIENT_SECRET
AZURE_AD_TENANT_ID=YOUR_TENANT_ID
SESSION_SECRET=SUPER_SECRET_STRING_USED_TO_SIGN_COOKIES
```

You also need [Redis](https://redis.io/) up and running locally, or modify `src/services/redis.ts` to connect to an external Redis.

Install and run the development server:

```bash
npm install
npm start
```

Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

> ## Remarks

Reviewer 2: Empty section?
